External Intercom

Saborknight opened the pull request:
**When merged this pull request will:**
- Let people ACE Interact with a vehicle to pickup/put away an intercom phone, allowing them to speak on the intercom of any ground vehicle
- Let people ACE Interact with a vehicle to connect wirelessly to the vehicle's intercom, allowing them to speak on the intercom of any ground or air vehicle, assuming that they're wearing compatible headgear
- Whenever a player wearing a wireless compatible headgear jumps out of an intercom capable vehicle's non-cargo seat, they are automatically connected to the intercom wirelessly, so that they can continue talking on the intercom without having to interact
- In all above cases, each of the two methods have their distinct maximum range of operation (Phone: 5m, Wireless: 15m). These values are customisable via CBA Settings

~~I'll be adding to this pull request with any suggested edits made, and by covering the below currently un-handled situations.~~ I think I'm done in terms of adding new features. ~~I'll also be making another pull request which will implement the physical telephone and rope, to add to immersion.~~ So it turns out I've done everything in one pull request. Oh well.

**Situations to handle correctly**
- [x] Player moves outside of the 5 meter limit -> Intercom phone disconnected
- [x] Another player tries to pickup an intercom phone whilst someone else has already picked up the same phone -> Display an "Intercom phone is in use" ACE Interaction instead of the "Pickup Intercom Phone" ACE interaction. Also, Intercom phone will not activate if already in use, in case the ACE interaction is bypassed
- [x] Player is killed, or disconnects from the server whilst connected to the intercom phone -> Intercom phone disconnected
- [x] Player gets into the same vehicle they are connected to with the intercom phone -> Intercom phone disconnected
- [x] Player gets into *any* vehicle whilst they are connected to the intercom phone -> will disconnect, though the distance checks should come into play if the vehicle the player is in, or the vehicle they are connected to move outside of the 5 meter radius
- [x] Player picks up another intercom phone whilst still being connected to a different intercom phone -> will disconnect old intercom phone
- [x] Player going unconscious (ACE) whilst being connected to the intercom phone -> will disconnect

**Todo for this initial implementation (AKA 1.0?)**
- [x] Add: option to disable external intercom, or only same side can connect
- [x] Add: Have wireless headgear reconnect automatically in case someone runs out of range by mistake. Make it so that they must disconnect via self interaction.
- [x] Allow wireless users to change intercom channel
- [x] Add: channel switch to phone panel, not self interact
- [x] Add: attribute that can be added to individual vehicle from Eden, Config or Zeus(?) to customise max range of phone and/or wireless
- [x] Fix: When jumping out of a vehicle, the intercom does stay connected wirelessly, but the icon doesn't appear?
- [ ] Fix: Rope breaking doesn't disconnect
- [x] Fix: AI jumping out will show indicator icon, even if Zeus is not on intercom
- [x] Fix: Now Headgear whitelist is suddenly not being transferred from server to client anymore...?
- [x] When being connected to wireless intercom, and changing the "Enable" setting to "Same side as Player", ropes attach to all players currently connected to the wireless intercom. So, one of the 2 players was given the rope and phone, but not actually assigned as phone operator... Meaning that player now can't put away the attached rope and phone
- [x] When one wireless user disconnects, all players lose the wireless indicator. This does not apply to someone when jumping in the vehicle (and wireless intercom disonnecting normally for them)
- [x] When a phone user is connected, and a crew member jumps out of the vehicle, the phone and cable disappear but icon remains, and intercom is still connected
- [x] When a wireless user takes helmet off, whilst another user is connected to wireless, it removes the phone and cable, but icon remains and intercom stays connected
- [x] Changing server settings don't apply globally
- [x] When no users are connected to the phone, but all are connected to the wireless, and the setting Enable is changed to something, then a random user (I assume first in the list) is actually connected to the phone instead of the wireless, and the other users are disconnected completely
- [x] When 2 users connected wirelessly, and setting gone from 1 to 2, locally each user is still considered connected, but the `TFAR_ExternalIntercomSpeakers` variable of the vehicle is not syncronised so we're each seeing an option to disconnect ourselves
- [x] When in crew seat, change to crew, then dismount. The channel will not remain on crew, it'll switch back to default
- [x] Connecting to wireless, getting into crew position, then getting out will duplicate person entry in `TFAR_ExternalIntercomSpeakers` of the vehicle
- [x] A player who isn't connected to intercom at all will have the icon appear randomly when another player jumps out of the vehicle
- [x] Modded Vehicles don't have the Intercom Phone interaction points?
- [x] Add entry to ACE Arsenal to indicate if a headgear is External Intercom Compatible
- [ ] ~~When all users are connnected to the wireless and the Enable setting is changed to 2, the second player in the list still has the "Disconnect Wireless" ACE interaction showing on the vehicle, even though they've been otherwise completely disconnected~~
- [ ] ~~A player (previously connected to wireless) isn't disconnected from the intercom when disconnecting from the server~~
- [ ] ~~Fix: Make sure `fnc_getIntercomChannel` sorts the array by `commander` > `gunner` > `driver` > `passengers` instead of the current sorting~~ Not so important if we just give the ability for the person to change channel themselves

> Note: Credits to Gary Sinise for donation of the handset model as well as corrections to certain calculations involved (He's the Maths voodoo man)

dedmen commented on addons/intercom_phone/ACEActions.hpp:
```diff
@@ -0,0 +1,31 @@
+            class TFAR_Intercom_Phone_Activate { \
+                displayName = CSTRING(PICKUP); \
+                distance = 4; \
+                condition = "isNil {_player getVariable ""TFAR_vehicleIDOverride""} && ([(typeOf _target), ""TFAR_hasIntercom"", 0] call TFAR_fnc_getVehicleConfigProperty) > 0 && isNil {_target getVariable [""TFAR_IntercomPhoneSpeaker"", nil]}"; \
```
```suggestion
                condition = "isNil {_player getVariable ""TFAR_vehicleIDOverride""} && ([(typeOf _target), ""TFAR_hasIntercom"", 0] call TFAR_fnc_getVehicleConfigProperty) > 0 && isNil {_target getVariable ""TFAR_IntercomPhoneSpeaker""}"; \
```

dedmen commented on addons/intercom_phone/CfgVehicles.hpp:
```diff
@@ -0,0 +1,16 @@
+    class LandVehicle;
+    ADD_ACTIONS(LandVehicle, Car)
+    ADD_ACTIONS(LandVehicle, Tank)
+    ADD_ACTIONS(LandVehicle, Motorcycle)
```
Do Motorcycle's ever have intercoms? :D 

I'd say that can be skipped, more important here would be helicopters I'd say

dedmen commented on addons/intercom_phone/RscTitles.hpp:
```diff
@@ -0,0 +1,28 @@
+                font = "PuristaMedium";
+                sizeEx = "(((((safezoneW / safezoneH) min 1.2) / 1.2) / 25) * 0.8)";
+                text=QPATHTOF(ui\tfar_intercom_phone.paa);
+                x="(profilenamespace getvariable [""IGUI_grid_TFAR_Volume_X"",  0.85 * safezoneW + safezoneX])";
```
Nice idea to reuse this.
But people can enable volume indicator to always show, that means these will then overlap eachother?
Maybe want to hide the speech volume indicator while you're in intercom

dedmen replied:
Actually you added a grid for intercom phone in config, but here you seem to use the volume one

Saborknight replied:
Honestly, I don't understand dialogues and, more specifically, how to position them at all... So was hoping someone would come round and comment on it like you did. So this grid, do I use the class name specified in https://github.com/michail-nikolaev/task-force-arma-3-radio/pull/1530/files#diff-5238f9c6d8630cb9ec6d642487487911310e6619735d338c5ea33bf702fa4aa2R46 ?

dedmen commented on addons/intercom_phone/functions/fnc_activate.sqf:
```diff
@@ -0,0 +1,73 @@
+};
+
+private _vehicleID = _vehicle getVariable ["TFAR_vehicleIDOverride", netid _vehicle];
+_vehicle setVariable ["TFAR_vehicleIDOverride", _vehicleID, true];
```
Is this needed? considering you just read that variable and know that (atleast locally) it already has that value?

dedmen commented on addons/intercom_phone/functions/fnc_activate.sqf:
```diff
@@ -0,0 +1,73 @@
+_player setVariable [format ["TFAR_IntercomSlot_%1",(netID _player)], _intercomSlot, true]; // Because TFAR checks the "vehicle", and thinks the unit is the vehicle when it's not in a vehicle, so we manually set it
+
+// Show indicator
+if (TFAR_oldVolumeHint) then {
```
Here we want to hide the volume indicator

dedmen commented on addons/intercom_phone/functions/fnc_activate.sqf:
```diff
@@ -0,0 +1,73 @@
+private _loopID = [_vehicle, _player] spawn {
+    params ["_vehicle", "_player"];
+
+    while {true} do {
```
I don't like spawned scripts. And also not while true loops. But this is appropriate here I guess.

1. The while loop should exit after deactivate is called
2. boundingBoxReal doesn't need to be re-calculated every time, that can be done once outside of the loop.

jokoho48 replied:
wouldnt something like this work:
```sqf
private _bbSize = (((boundingBoxReal _vehicle) select 2) / 2) + 5;
[{
    params ["_vehicle", "_player", "_bbSize"];
    (_vehicle distance _player) <= _bbSize
}, {
    params ["_vehicle", "_player"];
    [_vehicle, _player] call TFAR_intercom_phone_fnc_deactivate;
}, [_vehicle, _player, _bbSize]] call CBA_fnc_waitUntilAndExecute
```

Saborknight replied:
Thanks for that @jokoho48 ! I'll use your suggestion instead. It did a couple of errors in it but it ultimately works fine.

dedmen commented on addons/intercom_phone/functions/fnc_busyInteraction.sqf:
```diff
@@ -0,0 +1,32 @@
+#include "script_component.hpp"
+
+/*
+  Name: TFAR_fnc_intercomBusyIntercation
```
```suggestion
  Name: TFAR_fnc_intercomBusyInteraction
```

dedmen commented on addons/intercom_phone/functions/fnc_getIntercomChannel.sqf:
```diff
@@ -0,0 +1,28 @@
+
+// Get any occupants, from crew down to passengers
+private _crew = crew _vehicle;
+_vehicle getVariable [format ["TFAR_IntercomSlot_%1",(netId (_crew select 0))],
```
Is this variable set locally? I thought it was local only to remember what channel the player was last in when he left the vehicle.

dedmen replied:
> Do you mean that the variable for that crew member is local to that crew member's machine only?

Yes, atleast it doesn't make any sense for other players to know that, its only used for local cache afaik

Saborknight replied:
Eh.... Let me get back to you on this after some testing with my guys.

Saborknight replied:
Seems to work fine

Saborknight commented on addons/intercom_phone/functions/fnc_getIntercomChannel.sqf:
```diff
@@ -0,0 +1,28 @@
+params ["_vehicle"];
+
+// Get any occupants, from crew down to passengers
+private _crew = crew _vehicle;
```
It actually sucks that `crew` actually returns the list sorted by driver, gunner, commander and then passengers... when in actual fact it makes more sense to connect to the commander's channel, then the gunner's, then the driver's, then the passengers if the previous are not present...

dedmen commented on addons/external_intercom/baseHeadgear.hpp:
```diff
@@ -0,0 +1,20 @@
+private _baseHeadgear = [
```
This should be a config property, what should mods do to add their supported helmets to this?
I'd say make it config property, and scan once at preStart and save it in uiNamespace.

Or alternatively instead of scanning, could also make a global config class/array that mods can append to

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,148 @@
+_player setVariable [format ["TFAR_IntercomSlot_%1",(netId _player)], _intercomSlot, true]; // Because TFAR checks the "vehicle", and thinks the unit is the vehicle when it's not in a vehicle, so we manually set it
+
+// Show indicator
+if (TFAR_oldVolumeHint) then {
```
You don't need to use this. This was only specifically for Volume indicator (as some people didn't want the new on-screen indicator).

This is a new feature, you don't need to worry about backwards compat and people "preferring the old way"
I'd say take that out and only ever use the new indicator.

Or if you want to keep it, its unrelated to volume display so it should probably be separate setting then

Saborknight replied:
Oh hell no, I prefer the icon interface by far

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,148 @@
+    private _ropeID = ropeCreate [_vehicle, _interactionPointRelative vectorAdd [0,0.2,0], _handset, "plug", _intercomMaxRange]; // handset rope location[0,0.006,-0.7]
+    _vehicle setVariable ["TFAR_ExternalIntercomRopeIDs", [_ropeID, _handset], true];
+
+    _vehicle addEventHandler ["RopeBreak", {
```
Hah. This alone might be a sufficient "out of range" check :D

Saborknight replied:
I wish it was. The rope will stretch for ages rather than snap... which is kind of annoying. I wish we could set the elasticity of the rope to tell the engine how much give it should allow.

Actually... Does it exist?

Saborknight replied:
Once v2.06 of Arma is released we'll be able to make a custom rope to imitate wire, and change elasticity accordingly.

The code:
```c
class CfgVehicles {
    class Rope;
    class RopeRabbit : Rope {
        maxRelLenght = 1.1;
        maxExtraLenght = 20;
        springFactor = 0.5; // higher == less stretchy rope
        segmentType = "RopeSegmentRabbit";
        torqueFactor = 0.5;
        dampingFactor[] = {1.0,2.5,1.0};
        model = "\A3\Animals_F\rabbit\rabbit_F.p3d";
    };
};

class CfgNonAIVehicles {
    class RopeSegmentRabbit {
        scope=2;
        displayName="";
        simulation="ropesegment";
        autocenter=0;
        animated=0;
        model = "\A3\Animals_F\rabbit\rabbit_F.p3d";
    };
};
```
And to create it, just name the classname as an additional parameter to the usual ones for `ropeCreate`

nomisum replied:
Did you figure out how/when _ropeBreak_ works? Seems to work for direct connect to units for me but as soon as the ropeAttached object is attached itself, it wont. Still investigating if a workaround is possible.

Saborknight replied:
Nope, though if what you say is the case... maybe a base-game bug or something? Or maybe the eventhandler needs to be applied to something/one else? I would have thought the vehicle were enough, but maybe it needs to be applied to the phone too?

nomisum replied:
I am working on a phone mod and attach the rope from helperobject (heliclass) to player (lefthand selection) and with this setup ropeBreak works. you need to add the EH to the helperobject (heli class).

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,148 @@
+    _headgear = headgear _player;
+};
+
+[{
```
I'd suggest optimizing this by writing two conditions, one for wireless and one for not wireless.

You don't need modelToWorld (somewhat expensive) if its wireless, as center of mass is usually 0,0,0 so you can just use vehicle pos.

You also don't need to do the wireless check if not wireless.

You can do one check here, to save performance on every frame

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,148 @@
+    ["ace_unconscious", {
+        params ["_player", "_active"];
+        if (_active) then {
+            [_player getVariable "TFAR_ExternalIntercomVehicle", _player] call TFAR_external_intercom_fnc_disconnect;
```
Should it really disconnect if you go unconscious though?
If its wireless you still have your helmet/headset on even if you go unconscious

Saborknight replied:
Good point

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,148 @@
+
+_player setVariable ["TFAR_ExternalIntercomEHID",
+    ["ace_unconscious", {
+        params ["_player", "_active"];
```
should check player? Afaik ace_unconscious can also fire for local AI?

Saborknight replied:
I didn't think about that, good point

dedmen commented on addons/external_intercom/functions/fnc_disconnect.sqf:
```diff
@@ -0,0 +1,50 @@
+};
+(_externalIntercomSpeakers select 1) deleteAt ((_externalIntercomSpeakers select 1) find _player);
+
+_vehicle setVariable ["TFAR_ExternalIntercomSpeakers", _externalIntercomSpeakers, true];
```
potential race condition if two players disconnect at the same time in MP.
player1 might re-add player2 to the array even after player2 disconnected.
Not sure how to solve this, and also probably not that significant

Saborknight replied:
I'm not sure how to deal with the possible race condition either... maybe someone else will have an epiphane for us.

dedmen replied:
Let the server handle it.
Like with radio ID's. Tell the server "hey that happened" and the server takes care of checking and doing.
The server cannot do two things in parallel.

dedmen replied:
boop still open, though unlikely to happen. But maybe add a comment that thats possible and could potentially be fixed in the future

dedmen commented on addons/external_intercom/functions/fnc_getInteractionPoint.sqf:
```diff
@@ -0,0 +1,33 @@
+params ["_vehicle"];
+
+private _customPosition = getArray(configOf _vehicle);
```
does this make sense?
How can you call getArray on a vehicle class??

Saborknight replied:
Whoops

dedmen commented on addons/external_intercom/functions/fnc_getInteractionPoint.sqf:
```diff
@@ -0,0 +1,33 @@
+   _customPosition;
+};
+
+private [
```
Don't use private array, its a performance waste.
use the private keyword

dedmen commented on addons/external_intercom/functions/fnc_parseClassnamesInput.sqf:
```diff
@@ -0,0 +1,49 @@
+
+_parsedValue = _parsedValue arrayIntersect _parsedValue; // Retain only unique elements
+
+_parsedValue = _parsedValue apply {
```
Important! You are using `in` to check if headgear matches. `in` is case sensitive.
Meaning here you are relying on players to enter case-sensitively correct classnames. That is unlikely.


Instead of returning _x below, return `configName(configFile >> "CfgWeapons" >> _x)` that will get you correct configCase

Saborknight commented:
- [ ] Add option to disable this, or only same side can connect
- [ ] Idea: Have wireless headgear reconnect automatically in case someone runs out of range by mistake. Make it so that they must disconnect via self interaction.
Also, allow wireless users to change intercom channel
Add channel switch to phone panel, not self interact
Add attribute that can be added to individual vehicle from Eden, Config or Zeus(?) to customise max range of phone and/or wireless
Issue: When jumping out of a vehicle, the intercom does stay connected wirelessly, but the icon doesn't appear?
Issue: Rope breaking doesn't disconnect

Saborknight commented:
Some change since the last review has caused the headgear setting not to be transferred to clients anymore? When I fetch the setting from the server, it looks the way it should, but when I just try and use the global variable normally via just stating it, it comes up blank. I was under the impression that CBA would handle the transfer of the variable to any clients with the `onChange` parameter.... but it looks like I need to do it manually...

dedmen commented on addons/core/functions/fnc_preparePositionCoordinates.sqf:
```diff
@@ -27,9 +27,13 @@ private _isSpectating = _unit getVariable ["TFAR_forceSpectator",false];
 private _vehicle = "no"; //if (_isInVehicle) then {_unit call TFAR_fnc_vehicleId} else {"no"};
-if (!isNull (objectParent _unit) && !_isSpectating) then {//_isInVehicle
+private _isInVehicle = !isNull (objectParent _unit);
+if ((!(isNil {_unit getVariable "TFAR_vehicleIDOverride"}) || _isInVehicle) && !_isSpectating) then {//_isInVehicle
```
check _isInVehicle first, and do the isNil in lazy eval, isNil is somewhat a bit more expensive.

Maybe its cheaper to not use isNil?
maybe default value something like ctrlNull and then isEqualTo ctrlNull.
Would need to benchmark that.

This function runs for every unit for (almost) every frame, it needs to be top performing

dedmen commented on addons/core/functions/fnc_vehicleId.sqf:
```diff
@@ -19,7 +19,9 @@
 params ["_unit"];
 
-if (isNull (objectParent _unit) && isNil {_unit getVariable "TFAR_vehicleIDOverride"}) exitWith {"no"};//Unit is not in vehicle
+private _vehicleOverrideEnabled = !isNil {_unit getVariable "TFAR_vehicleIDOverride"};
```
this is the hot path.
The normal case is that its not enabled and the function immediately exits returning "no"

you invert the value with `!` twice in the hotpath, if you remove the ! on both you save two command calls

dedmen commented on addons/external_intercom/CustomInteractionPoints.hpp:
```diff
@@ -0,0 +1,97 @@
+
+class Tank_F;
+class MBT_02_base_F : Tank_F {
+    TFAR_ExternalIntercomInteractionPoint[] = {-1.26074,-4.75584,-0.938046};
```
would be cool if this can also support string memory point

Saborknight replied:
Good point!

dedmen commented on addons/external_intercom/WirelessHeadgear.hpp:
```diff
@@ -0,0 +1,19 @@
+class TFAR_External_Intercom_Wireless_Headgear {
+    // Vanilla + Apex + Contact
+    class H_HelmetCrew_B;
```
What are these classes here?
just use entry=1 like Joint Rails and some things in Arma do?

dedmen commented on addons/external_intercom/XEH_preInit.sqf:
```diff
@@ -0,0 +1,61 @@
+    1,
+    {
+        if (isServer) then {
+            TFAR_externalIntercomWirelessHeadgear = ((configFile >> "TFAR_External_Intercom_Wireless_Headgear") call BIS_fnc_getCfgSubClasses) + ([_this] call FUNC(parseClassnamesInput));
```
just directly https://community.bistudio.com/wiki/configProperties with "true" condition might be cheaper than calling the BIS func

dedmen commented on addons/external_intercom/XEH_preInit.sqf:
```diff
@@ -0,0 +1,61 @@
+    1,
+    {
+        _vehicle = ACE_Player getVariable "TFAR_ExternalIntercomVehicle";
+        if (_this > 0 && !isNil "_vehicle") then {
```
But if you disable it mid game, this doesn't disconnect you?

Saborknight replied:
Well.... in testing it has. Why wouldn't it? @dedmen

dedmen commented on addons/external_intercom/data/model.cfg:
```diff
@@ -0,0 +1,25 @@
+{
+	class Default
+	{
+		isDiscrete=1;
```
4 space indent. space before and after `=`

dedmen commented on addons/external_intercom/functions/fnc_addIntercomChannels.sqf:
```diff
@@ -0,0 +1,89 @@
+    [getText(_intercomRootClass >> _x >> "displayName"), _channelId];
+};
+
+if (_intercomChannels isEqualTo []) exitWith {false;};
```
```suggestion
if (_intercomChannels isEqualTo []) exitWith {false};
```

dedmen commented on addons/external_intercom/functions/fnc_addIntercomChannels.sqf:
```diff
@@ -0,0 +1,89 @@
+
+    _actions pushBack [
+        [
+            format["TFAR_ExternalIntercomChannel_%1", _intercomChannel],
```
```suggestion
            format ["TFAR_ExternalIntercomChannel_%1", _intercomChannel],
```

dedmen commented on addons/external_intercom/functions/fnc_addIntercomChannels.sqf:
```diff
@@ -0,0 +1,89 @@
+                _params params ["_intercomChannel", "_displayName"];
+
+                if (!isNil "_vehicle") then {
+                    _vehicle setVariable [format ["TFAR_IntercomSlot_%1",(netId _player)], _intercomChannel, true];
+                    _player setVariable [format ["TFAR_IntercomSlot_%1",(netId _player)], _intercomChannel, true];
```
```suggestion
                    _vehicle setVariable [format ["TFAR_IntercomSlot_%1", (netId _player)], _intercomChannel, true];
                    _player setVariable [format ["TFAR_IntercomSlot_%1", (netId _player)], _intercomChannel, true];
```

dedmen commented on addons/external_intercom/functions/fnc_addIntercomChannels.sqf:
```diff
@@ -0,0 +1,89 @@
+                params ["_vehicle", "_player", "_args", "_actionData"];
+                _args params ["_intercomChannel", "_displayName"];
+
+                _intercom = _player getVariable [format ['TFAR_IntercomSlot_%1',(netID _player)], -2];
```
```suggestion
                _intercom = _player getVariable [format ['TFAR_IntercomSlot_%1', (netID _player)], -2];
```

dedmen commented on addons/external_intercom/functions/fnc_addWirelessIntercomMenu.sqf:
```diff
@@ -0,0 +1,67 @@
+
+_vehicle = _player getVariable "TFAR_ExternalIntercomVehicle";
+
+if (isNil "_vehicle") exitWith {
+    [];
+};
```
```suggestion
if (isNil "_vehicle") exitWith {[]};
```

dedmen commented on addons/external_intercom/functions/fnc_addWirelessIntercomMenu.sqf:
```diff
@@ -0,0 +1,67 @@
+            [_player getVariable 'TFAR_ExternalIntercomVehicle', _player] call TFAR_external_intercom_fnc_disconnect;
+        },
+        {
+            params ["_vehicle", "_player"];
```
```suggestion
            params ["", "_player"];
```

dedmen commented on addons/external_intercom/functions/fnc_addWirelessIntercomMenu.sqf:
```diff
@@ -0,0 +1,67 @@
+        localize LSTRING(DISCONNECT_WIRELESS),
+        QPATHTOF(ui\tfar_ace_interaction_external_intercom_wireless_disconnect.paa),
+        {
+            params ["_vehicle", "_player"];
```
```suggestion
            params ["", "_player"];
```

dedmen commented on addons/external_intercom/functions/fnc_addWirelessIntercomMenu.sqf:
```diff
@@ -0,0 +1,67 @@
+        "",
+        {},
+        {
+            params ["_vehicle", "_player"];
```
```suggestion
            params ["", "_player"];
```

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,208 @@
+params ["_vehicle", "_player", "_actionParams"];
+_actionParams params [["_wireless", false, [false]]];
+
+if (!alive _vehicle) exitWith {
+    "no";
+};
```
```suggestion
if (!alive _vehicle) exitWith {"no"};
```

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,208 @@
+// Get intercom channel, now that we've signed on to the intercom
+private _intercomSlot = [_vehicle, _player] call TFAR_external_intercom_fnc_getIntercomChannel;
+
+_vehicle setVariable [format ["TFAR_IntercomSlot_%1",(netId _player)], _intercomSlot, true];
+// Because TFAR checks the "vehicle", and thinks the unit is the vehicle when it's not in a vehicle, so we manually set it
+_player setVariable [format ["TFAR_IntercomSlot_%1",(netId _player)], _intercomSlot, true];
```
```suggestion
_vehicle setVariable [format ["TFAR_IntercomSlot_%1", (netId _player)], _intercomSlot, true];
// Because TFAR checks the "vehicle", and thinks the unit is the vehicle when it's not in a vehicle, so we manually set it
_player setVariable [format ["TFAR_IntercomSlot_%1", (netId _player)], _intercomSlot, true];
```

dedmen commented on addons/external_intercom/CustomRope.hpp:
```diff
@@ -0,0 +1,23 @@
+class CfgVehicles {
+    class Rope;
+    class RopeSmallWire : Rope {
```
TFAR specific prefix to prevent mod conflicts

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,208 @@
+                    _vehicle removeEventHandler ["RopeBreak", _thisEventHandler];
+                };
+            }],
+            true
```
whaatt... is this true for? setVariable public with a EH ID? EH ID's are local only?

dedmen commented on addons/external_intercom/functions/fnc_connect.sqf:
```diff
@@ -0,0 +1,208 @@
+            true
+        ];
+        diag_log "Registered RopeBreak EH";
+        diag_log (_vehicle getVariable "TFAR_ExternalIntercomRopeEH");
```
missing cleanup.
You can use TRACE_ macros, and enable debug mode based on `__A3_DEBUG__` macro

dedmen commented on addons/external_intercom/functions/fnc_getInteractionPoint.sqf:
```diff
@@ -0,0 +1,42 @@
+private _customPosition = getArray((configOf _vehicle) >> "TFAR_ExternalIntercomInteractionPoint");
+
+// If there's a custom position, then don't bother calculating
+if !(_customPosition isEqualTo []) exitWith {
+   _customPosition;
+};
```
```suggestion
if !(_customPosition isEqualTo []) exitWith {_customPosition};
```

dedmen commented on addons/external_intercom/functions/fnc_getIntercomChannel.sqf:
```diff
@@ -0,0 +1,31 @@
+params ["_vehicle", "_player"];
+
+// Get any occupants, from crew down to passengers
+private _crew = crew _vehicle;
+_vehicle getVariable [format ["TFAR_IntercomSlot_%1", netId _player],
+    _vehicle getVariable [format ["TFAR_IntercomSlot_%1",(netId (_crew select 0))],
```
```suggestion
private _crew = crew _vehicle;
_vehicle getVariable [format ["TFAR_IntercomSlot_%1", netId _player],
    _vehicle getVariable [format ["TFAR_IntercomSlot_%1", (netId (_crew select 0))],
```

dedmen commented on addons/external_intercom/functions/fnc_parseClassnamesInput.sqf:
```diff
@@ -0,0 +1,40 @@
+];
+
+private _parsedValue = (_valueString splitString ",") apply {
+    if !(_x isEqualType "STRING") then {_x = format["%1", _x]};
```
```suggestion
    if !(_x isEqualType "") then {_x = format["%1", _x]};
```
maybe a bit less readable, but.. more efficient? extremely tiny bit? Doesn't really matter

dedmen commented on addons/external_intercom/functions/fnc_parseClassnamesInput.sqf:
```diff
@@ -0,0 +1,40 @@
+
+private _parsedValue = (_valueString splitString ",") apply {
+    if !(_x isEqualType "STRING") then {_x = format["%1", _x]};
+    [_x] call BIS_fnc_filterString;
```
```suggestion
    [_x] call BIS_fnc_filterString
```
if you intend to return this value leave away semicolon on return values

Saborknight replied:
And does this apply to the end of file lines too? End of functions, where it outputs the function return?

dedmen commented on addons/external_intercom/functions/fnc_parseClassnamesInput.sqf:
```diff
@@ -0,0 +1,40 @@
+
+_parsedValue = _parsedValue arrayIntersect _parsedValue; // Retain only unique elements
+
+_parsedValue = _parsedValue apply {
```
```suggestion
_parsedValue = _parsedValue apply { // convert to config case
```

dedmen commented on addons/external_intercom/functions/fnc_parseClassnamesInput.sqf:
```diff
@@ -0,0 +1,40 @@
+    } else {""};
+};
+
+_parsedValue = _parsedValue select {!(_x isEqualTo "")}; // Retain only non-null elements
```
```suggestion
_parsedValue = _parsedValue - [""]; // Retain only non-null elements
```

dedmen commented on addons/external_intercom/functions/fnc_settingForceAlphaNumericArray.sqf:
```diff
@@ -0,0 +1,29 @@
+if (_input isEqualTo "") exitWith {""};
+// keep the last sign, if you want to type a new value
+private _lastSign = _input select [count _input - 1];
+_lastSign = if !((_lastSign isEqualTo ",")||{_lastSign isEqualTo " "}) then {""};
```
```suggestion
_lastSign = if !((_lastSign isEqualTo ",") || {_lastSign isEqualTo " "}) then {""};
```

dedmen commented on addons/external_intercom/script_component.hpp:
```diff
@@ -0,0 +1,18 @@
+#define COMPONENT external_intercom
+#include "\z\tfar\addons\core\script_mod.hpp"
+
+#ifdef DEBUG_ENABLED_EXTERNAL_INTERCOM
```
In Arma 2.06 you can set preprocessor macros via start parameter
https://community.bistudio.com/wiki/Arma_3:_Startup_Parameters
preprocDefine
could use this here to dynamically enable intercom debugging

Saborknight replied:
Not too sure how that all works? This was just a copy/paste from one of the other addons

Saborknight replied:
Oh, I see what you mean. You're saying I can set this variable to true in my launcher for additional logging. Thanks for the heads up!

dedmen commented on addons/external_intercom/ui/tfar_ace_interaction_external_intercom_phone.svg:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
```
These don't need to be inside a pbo, waste a bit of space
We have other source files in https://github.com/michail-nikolaev/task-force-arma-3-radio/tree/master/extra
but this is small enough to not matter

Saborknight commented on addons/external_intercom/ACEActions.hpp:
```diff
@@ -0,0 +1,78 @@
+#define ADD_PHONE_ACTIONS \
+class TFAR_External_Intercom_Phone_Base { \
```
I see what you're saying and that does make sense, but from the player's perspective, having to go through an extra menu item when interacting with a vehicle that could potentially be moving at the time would probably be difficult to do. Even if not moving, it's still an extra step that UX wise doesn't make sense....
Is there a way perhaps to still make it more efficient without making children? Or do some quick replacement maybe?.... maybe I could add the interactions via script instead, and have an overall check there to see if the vehicle even has intercom. What do you think @dedmen ?

Saborknight commented on addons/external_intercom/CfgVehicles.hpp:
```diff
@@ -1,41 +1,17 @@
-    }; \
-};
+#define LAND_INTERACTIONS \
+init = QUOTE(call FUNC(addWirelessInteractions);call FUNC(addPhoneInteractions));
```
Ah well, in that case I'll move to that instead. Thanks!

Saborknight replied:
> 

So I've tested it, and it doesn't seem to execute on an object unless I actually mount into it. So, this (using the `newControllableObject` eventhandler) isn't going to work out.

nomisum commented:
eagerly waiting for this, want to use that rope and model as a dependency for a telephone mod :) 

love the external intercom feature too.

Saborknight commented:
Is this waiting on anything else? I can't really make any more changes past the 26th as I'm hoping to be in the military and therefore won't be available at that point. So, anything that you guys need from me, please let me know beforehand. Thanks

dedmen commented:
Ugh...
Many build errors.
ACEActions.hpp has two TFAR_Radios classes, that can't be.
CustomInteractionPoints.hpp has multiple base class declarations for B_APC_Tracked_01_base_F, that also can't be

dedmen commented:
```
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_SOMEONE_USING_PHONE"><Original>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_SOMEONE_USING_PHONE"><English>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_SOMEONE_USING_PHONE"><Czech>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_SOMEONE_USING_PHONE"><Russian>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_SOMEONE_USING_PHONE"><Italian>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_HEADGEAR_NOT_WIRELESS_CAPABLE"><Original>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_HEADGEAR_NOT_WIRELESS_CAPABLE"><English>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_HEADGEAR_NOT_WIRELESS_CAPABLE"><Czech>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_HEADGEAR_NOT_WIRELESS_CAPABLE"><Russian>
18:51:00 Unexpected stringtable format inside <Text ID="STR_TFAR_EXTERNAL_INTERCOM_HEADGEAR_NOT_WIRELESS_CAPABLE"><Italian>
```


😢
